python/tank_vendor/shotgun_deploy/descriptor_io/dev.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+# 
+# CONFIDENTIAL AND PROPRIETARY
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+from .path import IODescriptorPath
+from ..errors import ShotgunDeployError
+
+from .. import util
+log = util.get_shotgun_deploy_logger()
+
+class IODescriptorDev(IODescriptorPath):
+    """
+    Represents a local dev item. This item is never downloaded
+    into the local storage, you interact with it directly::
+
+        {"type": "dev", "path": "/path/to/app"}
+
+    Optional parameters are possible::
+
+        {"type": "dev", "path": "/path/to/app", "name": "my-app"}
+
+        {"type": "dev",
+         "linux_path": "/path/to/app",
+         "windows_path": "d:\foo\bar",
+         "mac_path": "/path/to/app" }
+
+    String urls are on the following form::
+
+        sgtk:dev:[name]:local_path
+        sgtk:dev3:[name]:win_path:linux_path:mac_path
+
+        sgtk:dev:my-app:/tmp/foo/bar
+        sgtk:dev3::c%3A%0Coo%08ar:/tmp/foo/bar:
+
+    Name is optional and if not specified will be determined based on folder path.
+    If name is not specified and path is /tmp/foo/bar, the name will set to 'bar'
+    """
+
+    def __init__(self, location_dict):
+        """
+        Constructor
+
+        :param location_dict: Location dictionary describing the bundle
+        :return: Descriptor instance
+        """
+        super(IODescriptorDev, self).__init__(location_dict)
+
+    @classmethod
+    def dict_from_uri(cls, uri):
+        """
+        Given a location uri, return a location dict
+
+        :param uri: Location uri string
+        :return: Location dictionary
+        """
+        # sgtk:dev:[name]:local_path
+        # sgtk:dev3:[name]:win_path:linux_path:mac_path
+        #
+        # Examples:
+        # sgtk:dev:my-app:/tmp/foo/bar
+        # sgtk:dev3::c%3A%0Coo%08ar:/tmp/foo/bar:
+
+        # explode into dictionary
+        location_dict = None
+        try:
+            location_dict = cls._explode_uri(
+                uri,
+                "dev",
+                ["name", "path"]
+            )
+        except ShotgunDeployError:
+            # probably because it's a dev3, not a dev
+            pass
+
+        try:
+            location_dict = cls._explode_uri(
+                uri,
+                "dev3",
+                ["name", "windows_path", "linux_path", "mac_path"]
```

Comment:
great! current format is placeholder, i didn't even try to come up with someting sensbible, was more trying to get the structure and high level concepts right for bootstrap. @robblau suggested `sgtk:descriptor:app_store?name=foo&version=bar` which I really like. I'll put your comments on our hot topics doc for discussion :)